### PR TITLE
Fix arithmetic errors with timedelta64 dtypes

### PR DESCRIPTION
**Pull request by jbrockmendel**
Kind of surprising, but I don't see any Issues for these bugs.  Might be because many of them were found after parametrizing arithmetic tests, so instead of Issues we had xfails.

xref #20088 
xref #22163

- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -576,7 +576,10 @@ Datetimelike
 - Bug in :class:`DataFrame` with mixed dtypes including ``datetime64[ns]`` incorrectly raising ``TypeError`` on equality comparisons (:issue:`13128`,:issue:`22163`)
 - Bug in :meth:`DataFrame.eq` comparison against ``NaT`` incorrectly returning ``True`` or ``NaN`` (:issue:`15697`,:issue:`22163`)
 - Bug in :class:`DataFrame` with ``timedelta64[ns]`` dtype division by ``Timedelta``-like scalar incorrectly returning ``timedelta64[ns]`` dtype instead of ``float64`` dtype (:issue:`20088`,:issue:`22163`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you split out to a timedelta bug section (as this is getting kind of long)?

can be in a future PR

File: pandas/core/indexes/range.py [hunk 2]
```diff
@@ -596,6 +597,9 @@ def _evaluate_numeric_binop(self, other):
                     # GH#19333 is_integer evaluated True on timedelta64,
                     # so we need to catch these explicitly
                     return op(self._int64index, other)
+                elif is_timedelta64_dtype(other):
```
Comment (jreback) [inline pandas/core/indexes/range.py]: doesn't the fall thru raise TypeError?

File: pandas/core/ops.py [hunk 3]
```diff
@@ -1244,6 +1244,32 @@ def wrapper(left, right):
                                     index=left.index, name=res_name,
                                     dtype=result.dtype)
 
+        elif is_timedelta64_dtype(right) and not is_scalar(right):
+            # i.e. exclude np.timedelta64 object
+            # GH#22390 Unfortunately we need to special-case right-hand
+            # timedelta64 dtypes because numpy casts integer dtypes to
+            # timedelta64 when operating with timedelta64
+            if isinstance(right, np.ndarray):
+                # upcast to TimedeltaIndex before dispatching
```
Comment (jreback) [inline pandas/core/ops.py]: uhh this seems like a lot of special casing, any way to make this simpler

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt (hunk 1 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v0.24.0.txt]: Sounds good

File: pandas/core/ops.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: The only simplification I considered but didn't use was something like:

```
def maybe_upcast_for_op(obj):
     if type(obj) is timedelta:
        return pd.Timedelta(obj)
    if isinstance(obj, np.ndarray) and is_timedelta64_dtype(obj):
        return pd.TimedeltaIndex(obj)
    return obj
```

which would go right after `get_op_result_name`.

**Comment by jbrockmendel** [comment]
This will probably need rebasing after #22350 goes through

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/range.py]: I don’t understand the question

File: pandas/core/ops.py [hunk 4]
```diff
@@ -1244,6 +1244,32 @@ def wrapper(left, right):
...
+
+            # Note: we cannot use dispatch_to_index_op because
+            # that may incorrectly raise TypeError when we
+            # should get NullFrequencyError
+            result = op(pd.Index(left), right)
+            return construct_result(left, result,
+                                    index=left.index, name=res_name,
+                                    dtype=result.dtype)
+
+        elif type(right) is datetime.timedelta:
```
Comment (jreback) [inline pandas/core/ops.py]: this case you can include above (before what you added), no?

```
if isintance(right, (datetime.timedelta, Timedelta)):
  right = Timedelta(right)
....
```

File: pandas/core/ops.py (hunk 3 above)
Comment (jreback) [reply pandas/core/ops.py]: maybe re-route the scalars to a separate path. This is getting way special casey here.

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexes/range.py]: your comment is at odds with the checking i think. pls revise comments and/or checks.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/range.py]: The comment below this line `# must be an np.ndarray; GH#22390` along with the check above `elif is_timedelta64_dtype(other)` is just saying this is an `ndarray["timedelta64['ns']"]`.  I don't think these are at odds.

File: pandas/core/ops.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: I'll implement the maybe_upcast idea and see if that is prettier

File: pandas/core/ops.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: The version code above casts to `TimedeltaIndex`, _not_ `Timedelta`.

File: doc/source/whatsnew/v0.24.0.txt (hunk 1 above)
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: I *think* we now have a separate section?

**Review by jreback — APPROVED** [review]
minor doc-comment. rebase and ping on green.

**Comment by jbrockmendel** [comment]
ping

**Comment by jreback** [comment]
thanks! nice cleaning

**Pull request merged**